File: alg-teachers-support-sorting/3 var/2_task.py
n = int(input())  # Считываем размерность массива
arr = list(map(int, input().split()))  # Считываем элементы массива и преобразуем их в список целых чисел

# Считаем количество повторений для каждого элемента
frequencies = {}
for num in arr:
    frequencies[num] = frequencies.get(num, 0) + 1

# Сортируем массив по количеству повторений и, в случае равенства, по возрастанию
for i in range(len(arr)-1):
    for j in range(len(arr)-i-1):
        if frequencies[arr[j]] == frequencies[arr[j+1]]:
            if arr[j]>arr[j+1]:
                arr[j], arr[j+1] = arr[j+1], arr[j]
        elif frequencies[arr[j]] < frequencies[arr[j+1]]:
            arr[j], arr[j+1] = arr[j+1], arr[j]
sorted_arr = arr

for i in range(len(arr)):
    if i!=0:
        if not sorted_arr[i]==sorted_arr[i-1]:
            print(sorted_arr[i], frequencies[sorted_arr[i]])
    else:
        print(sorted_arr[i], frequencies[sorted_arr[i]])
print("Answer:", ' '.join(map(str, sorted_arr)))

File: alg-teachers-support-sorting/3 var/2_task.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

bool compare(const std::pair<int, int>& a, const std::pair<int, int>& b) {
    if (a.second != b.second) {
        return a.second < b.second; // Сортировка по убыванию количества повторений
    }
    return a.first > b.first; // Сортировка по возрастанию элементов в случае равенства количества повторений
}

void sortByFrequency(int arr[], int n) {
    std::unordered_map<int, int> frequencyMap;
    for (int i = 0; i < n; i++) {
        frequencyMap[arr[i]]++;
    }

    
    std::vector<std::pair<int, int>> frequencyPairs;
    for (const auto& pair : frequencyMap) {
        frequencyPairs.push_back(pair);
    }

    for(int i=1;i<frequencyPairs.size();i++)     
	   for(int j=i;j>0 && compare(frequencyPairs[j-1],frequencyPairs[j]); j--) {
			std::swap(frequencyPairs[j-1],frequencyPairs[j]);
       }

    for (const auto& pair : frequencyPairs) {
        std::cout << pair.first << " " << pair.second << std::endl;
    }

    int index = 0;
    for (const auto& pair : frequencyPairs) {
        int num = pair.first;
        int count = pair.second;
        for (int i = 0; i < count; i++) {
            arr[index++] = num;
        }
    }
}

void printArray(int arr[], int n) {
    for (int i = 0; i < n; i++) {
        if (i == n-1) std::cout << arr[i] << std::endl;
        else std::cout << arr[i] << " ";
    }
}

int main() {
    int n;
    std::cin >> n;
    int arr[n];
    for (int i = 0; i<n; i++) {
        std::cin >> arr[i];
    }


    sortByFrequency(arr, n);

    std::cout << "Answer: ";
    printArray(arr, n);

    return 0;
}